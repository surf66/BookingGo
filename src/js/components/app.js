import React from 'react';
import Form from './form';

export default class App extends React.Component {
  render() {
    return(
      <div className="container">
        <Form />
      </div>
    );
  }
}