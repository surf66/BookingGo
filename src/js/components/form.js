import React from 'react';
import SearchField from './search-field';
import SuggestionList from './suggestion-list';

const Form = () => {
  return(
    <section>
      <form className="pickup-form">
        <h1>Let's find your ideal car</h1>
        <SearchField />
        <SuggestionList />
      </form>
    </section>
  );
}

export default Form;