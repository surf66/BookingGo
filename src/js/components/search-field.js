import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { debounce } from 'throttle-debounce';
import { SearchTermActions } from '../actions/index';

function mapDispatchToProps(dispatch) {
  return ({
    setSearchTerm: bindActionCreators(SearchTermActions.setSearchTerm, dispatch)
  })
}

export class SearchField extends React.Component {

  constructor() {
    super();

    this.handleChange = this.handleChange.bind(this);
    this.storeSearchTerm = debounce(1000, this.storeSearchTerm);
  }

  render() {
    return(
      <div className="field-container">
        <label htmlFor="pickup-location">Pick-up Location</label>
        <input type="text" id="pickup-location" name="pickup-location" placeholder="city, airport, station, region, district..." onChange={this.handleChange} />
      </div>
    );
  }

  handleChange(event) {
    let searchTerm = event.target.value;
    this.storeSearchTerm(searchTerm);
  }

  storeSearchTerm(searchTerm) {
    this.props.setSearchTerm(searchTerm);
  }
}

export default connect(null, mapDispatchToProps)(SearchField);