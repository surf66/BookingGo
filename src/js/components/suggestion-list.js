import React from 'react';
import { connect } from 'react-redux';
import SuggestionsService from '../services/suggestions-service';

function mapStateToProps(state) {
  return {
    searchTerm: state.searchTerm,
  }
}

export class SuggestionList extends React.Component {

  constructor() {
    super();

    this.state = {
      suggestions: []
    }

    this.getSuggestions = this.getSuggestions.bind(this);
    this.highlightSearchTerm = this.highlightSearchTerm.bind(this);
  }
  
  componentWillReceiveProps(newProps) {
    this.getSuggestions(newProps.searchTerm)
  }

  render() {
    return(
      <div className={this.state.suggestions.length > 0 ? 'list open' : 'list'}>
        {this.state.suggestions.map((suggestion, index) => {
          var suggestionName = this.highlightSearchTerm(suggestion.name, this.props.searchTerm);
          return <div className="item" key={index}>
            <p className="title" dangerouslySetInnerHTML={{__html: suggestionName}}></p>
            {suggestion.region && suggestion.country && <p className="description">{suggestion.region}, {suggestion.country}</p>}
          </div>;
        })}
      </div>
    );
  }

  async getSuggestions(searchTerm) {
    if(searchTerm.length > 1) {
      let suggestions = await SuggestionsService.getSuggestions(searchTerm);
      this.setState({ suggestions });
      return;
    }

    this.setState({ suggestions: [] });
    return;
  }

  highlightSearchTerm(suggestion, searchTerm) {
    return suggestion.toLowerCase().replace(searchTerm, `<span>${searchTerm}</span>`);
  }
}

export default connect(mapStateToProps, null)(SuggestionList);