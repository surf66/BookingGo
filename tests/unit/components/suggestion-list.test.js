import React from 'react';
import { SuggestionList } from '../../../src/js/components/suggestion-list';
import SuggestionsService from '../../../src/js/services/suggestions-service';

const mockSuggestions = [
  { name: 'Manchester Airport', city: 'Manchester', country: 'UK' },
  { name: 'Wigan Pier', city: 'Greater Manchester', country: 'UK' }
];

test('should match snapshot when no suggestions provided', () => {
  const component = mount(<SuggestionList />);

  expect(component).toMatchSnapshot();
});

test('should match snapshot when suggestions provided', () => {
  const component = mount(<SuggestionList searchTerm="Manchester" />);
  component.setState({ suggestions: mockSuggestions });
  component.update();

  expect(component).toMatchSnapshot();
});

test('should call SuggestionService when getSuggestions is called', async () => {
  spyOn(SuggestionsService, 'getSuggestions').and.returnValue(mockSuggestions);
  let isRequestingMock = jasmine.createSpy();
  
  const mockSearchTerm = 'testsearchterm';
  const component = shallow(<SuggestionList isRequesting={isRequestingMock} searchTerm="Manchester" />);

  await component.instance().getSuggestions(mockSearchTerm);

  expect(isRequestingMock).toHaveBeenCalledWith(true);
  expect(SuggestionsService.getSuggestions).toHaveBeenCalledWith(mockSearchTerm);
});

test('should set suggestions in state when getSuggestions is called and suggestions are returned', async () => {
  SuggestionList.prototype.setState = jasmine.createSpy();
  let isRequestingMock = jasmine.createSpy();
  spyOn(SuggestionsService, 'getSuggestions').and.returnValue(mockSuggestions);
  
  const mockSearchTerm = 'testsearchterm';
  const component = shallow(<SuggestionList isRequesting={isRequestingMock} />);

  await component.instance().getSuggestions(mockSearchTerm);

  expect(isRequestingMock).toHaveBeenCalledWith(true);
  expect(SuggestionList.prototype.setState).toHaveBeenCalledWith({ suggestions: mockSuggestions });
});

test('should set suggestions in state to an empty array when searchTerm is less than 1', async () => {
  SuggestionList.prototype.setState = jasmine.createSpy();
  const mockSearchTerm = 't';
  const component = shallow(<SuggestionList />);

  await component.instance().getSuggestions(mockSearchTerm);

  expect(SuggestionList.prototype.setState).toHaveBeenCalledWith({ suggestions: [] });
});

test('should call getSuggestions when component receives updated props', () => {
  SuggestionList.prototype.getSuggestions = jasmine.createSpy();
  const component = mount(<SuggestionList />);
  const mockNewProps = { searchTerm: 'testsearchterm' };

  component.instance().componentWillReceiveProps(mockNewProps);

  expect(SuggestionList.prototype.getSuggestions).toHaveBeenCalledWith(mockNewProps.searchTerm);
});