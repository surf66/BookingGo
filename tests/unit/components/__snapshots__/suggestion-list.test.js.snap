// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot when no suggestions provided 1`] = `
<SuggestionList>
  <div
    className="list"
  />
</SuggestionList>
`;

exports[`should match snapshot when suggestions provided 1`] = `
<SuggestionList>
  <div
    className="list open"
  >
    <div
      className="item"
      key="0"
    >
      <p
        className="title"
        dangerouslySetInnerHTML={
          Object {
            "__html": "manchester airport",
          }
        }
      />
    </div>
    <div
      className="item"
      key="1"
    >
      <p
        className="title"
        dangerouslySetInnerHTML={
          Object {
            "__html": "wigan pier",
          }
        }
      />
    </div>
  </div>
</SuggestionList>
`;
